CMSgov/easi-app
# [NOREF] update pre commit hooks

# NOREF 

## Changes and Description

- Updates the versions of many pre-commit hooks
- Adds a readable `name` to each hook
- Fix some scripts that were running on all files being executed multiple times

This PR started as a fix for `check-code-generation` sometimes failing in CI, which likely turned out to be because that script does not take filenames as an argument, but instead runs on _all_ files, but pre-commit defaults to assuming that a script should be run once for each file.

Using `pass_filenames` fixes this for scripts like this, so it was introduced on a few of them. Others, where possible (like `eslint`) were refactored to instead take filenames as positional arguments so that files could be linted individually, rather than having to lint everything at once (or hard-coding git commands, which makes things like `pre-commit run --all-files` not actually work at all in CI)

## Comparison
### Before
```
$ time scripts/dev lint
pre-commit run -a
Enforce branch naming pattern............................................Failed
- hook id: no-commit-to-branch
- exit code: 1
Shell Syntax Check.......................................................Passed
check yaml...............................................................Passed
detect private key.......................................................Passed
Generate a Docs Index....................................................Passed
Run spellcheck...........................................................Passed
go import/fmt............................................................Passed
golangci-lint............................................................Passed
Dockerfile linter........................................................Passed
eslint...................................................................Passed
graphql-schema...........................................................Passed
check-code-generation....................................................Passed
check-intake-schema-generation...........................................Passed
preprocess-intake-swagger................................................Passed
Git Secrets..............................................................Passed
Add newlines to end of files.............................................Passed
Command failed with status (1): [pre-commit run -a...]

real    0m34.312s
user    2m15.661s
sys     1m52.340s
```
### After
```
$ time scripts/dev lint
pre-commit run -a
Enforce branch naming pattern............................................Passed
Check YAML formatting....................................................Passed
Check for private keys...................................................Passed
Add newlines to ends of files............................................Passed
Lint shell files.........................................................Passed
Spellcheck Markdown files................................................Passed
Run "goimports" on go files..............................................Passed
Lint go files............................................................Passed
Lint Dockerfile..........................................................Passed
Run eslint...............................................................Passed
Lint GraphQL schema......................................................Passed
Run GraphQL code generation..............................................Passed
Generate CEDAR Intake schema.............................................Passed
Generate CEDAR Intake client.............................................Passed
Check for committed secrets..............................................Passed

real    0m19.704s
user    1m9.379s
sys     0m13.884s
```

This is a ~44% decrease in start-to-finish time of the full lint! 🎉 
Executions in CI are also faster by using a new version of the `pre-commit/action` github action, which includes default pre-commit action caching!

## How to test this change

- Pull this branch down
- Make changes that you think should break the linting process
- Run `pre-commit run --all-files`

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
